[DAG] Use known-bits when creating umulh/smulh.

davemgreen opened the pull request:
This extends the creation of umulh/smulh instructions to handle cases where one operand is a zext/sext and the other has enough known-zero or sign bits to create a mulh. This can be useful when one of the operands is hoisted out of a loop.

RKSimon reviewed (commented):
CodeGen/X86/combine-pmuldq.ll change?

davemgreen commented:
> CodeGen/X86/combine-pmuldq.ll change?

Sorry - I messed up my testing the first time. As far as I can tell there is some custom lowering for mulh in X86 which adds some shuffles, and in the process produces worse instructions. IIUC it ends up with `vshuf(vperm(vshuf(vbroadcast(...)))`, but the broadcast is in a separate BB. We would usually solve / work around this on AArch64 by sinking the broadcast into the loop in CGP. It looks like the shuffling is causing more instructions in some other places too. Do you have a preferred way of handling this in X86?

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -10821,18 +10825,17 @@ static SDValue combineShiftToMULH(SDNode *N, const SDLoc &DL, SelectionDAG &DAG,
+        DAG.computeKnownBits(RightOp).countMinLeadingZeros() >= NarrowVTSize) {
+      MulhRightOp = DAG.getNode(ISD::TRUNCATE, DL, NarrowVT, RightOp);
+    } else if (IsSignExt && ShiftOperand.hasOneUse() &&
+               DAG.ComputeNumSignBits(RightOp) > NarrowVTSize) {
```
As the mul needs twice the number of bits it was assuming they were the same. There is a check below that `WideVT.getScalarSizeInBits() == 2 * NarrowVTSize`.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -10841,18 +10845,17 @@ static SDValue combineShiftToMULH(SDNode *N, const SDLoc &DL, SelectionDAG &DAG,
-                              ? Constant->getAPIntValue().getSignificantBits()
-                              : Constant->getAPIntValue().getActiveBits();
-    if (ActiveBits > NarrowVTSize)
+  if (LeftOp.getOpcode() != RightOp.getOpcode()) {
```
If I understand correctly, this patch amounts to replacing the special case of constants with value tracking, then why is the swap above necessary?

davemgreen replied:
The old code assumed that the constant had been canonicalized to the RHS. We now are able to detect and optimize patterns like `srl(mul(and(b), zext(a)))`, converting them to `umulh(trunc(a), b)`. The swap is to allow the known-bits `and` operand to be either way around.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -10809,6 +10809,10 @@ static SDValue combineShiftToMULH(SDNode *N, const SDLoc &DL, SelectionDAG &DAG,
   SDValue RightOp = ShiftOperand.getOperand(1);
 
+  if (LeftOp.getOpcode() != ISD::SIGN_EXTEND &&
+      LeftOp.getOpcode() != ISD::ZERO_EXTEND)
```
What about any_extend?

RKSimon reviewed (approved):
LGTM - I'll continue #188480 as a follow up - sorry for the delay

davemgreen committed:
[DAG] Use known-bits when creating umulh/smulh.

This extends the creation of umulh/smulh instructions to handle cases where one
operand is a zext/sext and the other has enough known-zero or sign bits to
create a mulh. This can be useful when one of the operands is hoisted out of a
loop.
